### PR TITLE
feat(mcp): read-only MCP diagnostic server

### LINKED ISSUE #180: Provide an official read-only MCP server for pgbus diagnostics (secure, structured queue/worker/health access)
## Summary

Provide an **official, read-only MCP (Model Context Protocol) server** for pgbus that exposes queue/worker/job/health state as structured tools, so an AI agent (or any MCP client) can diagnose pgbus problems directly instead of hand-writing `pgmq` / `pg_stat_activity` SQL against the production database.

Motivation: the silent-worker-wedge class of incident (#179, previously #174) is invisible on every existing pgbus health surface — the dashboard shows workers **Healthy** while the claim loop is dead and queues back up with `read_ct = 0`. Diagnosing it required manually querying the pgmq queue tables, `pgbus_queue_states`, `pg_locks`, and `pg_stat_activity`. An MCP that surfaces those same truths as safe, structured tools would turn a multi-hour forensic session into a single query, and let agents detect/triage the condition automatically.

## Why this is low-risk to build

pgbus already has the entire read layer this needs: `lib/pgbus/web/data_source.rb` powers the dashboard with methods like `queues_with_metrics`, `queue_detail`, `processes`, `job_locks`, `queue_health_stats`, `job_throughput`, `job_status_counts`, `dlq_messages`, `recurring_tasks`, `outbox_stats`, `batches`. The MCP server is largely a **thin, read-only adapter** over this existing, battle-tested query surface — not new database access.

## Proposed tools (read-only)

Map the most diagnostic `DataSource` methods to MCP tools. Suggested set:

- `pgbus_queues` — all queues with depth, visible count, oldest-message age, throughput (from `queues_with_metrics`).
- `pgbus_queue_detail(name)` — per-queue detail incl. paused state and health stats (`queue_detail`, `queue_paused?`, `queue_health_detail`).
- `pgbus_processes` — every process (worker/consumer/dispatcher/scheduler/supervisor) with kind, pid, heartbeat age, and — once #179 lands — claim-loop-progress age and Stalled flag.
- `pgbus_health` — top-level health summary: any queue with `visible > 0` and no `read_ct` progress while a subscribing worker has idle capacity (the single signal that catches the silent-wedge class). Returns a clear OK / DEGRADED / STALLED verdict.
- `pgbus_jobs(queue, page)` / `pgbus_job_detail(queue, msg_id)` — inspect enqueued jobs incl. `read_ct`, `vt`, enqueued_at.
- `pgbus_dlq` / `pgbus_dlq_detail(msg_id)` — dead-letter inspection.
- `pgbus_locks` — job/advisory locks (the leaked-uniqueness-lock diagnostic).
- `pgbus_throughput(minutes)` / `pgbus_stats(minutes)` — recent throughput and status counts.
- `pgbus_recurring` — recurring task schedule + last-run times.

These cover the exact queries needed to diagnose #179/#174 without raw SQL: "are queues backed up?", "is read_ct advancing?", "are workers heart-beating but not claiming?", "is anything paused or locked?".

## Security requirements (the "secure way" the title asks for)

The MCP must be safe to expose against a production datastore:

- **Read-only by default.** No tool mutates state. Any future write/admin tools (retry job, purge DLQ, resume queue, recycle worker) must be **separate, explicitly opt-in, and off by default** — never bundled into the diagnostic surface.
- **No raw SQL passthrough.** Tools are a fixed, curated set mapped to `DataSource` methods. Do not expose an arbitrary-query tool.
- **No sensitive payload leakage.** Job arguments / message bodies can contain PII or secrets. Diagnostic tools should return metadata (counts, ages, ids, read_ct, vt, queue names, job_class) by default and **redact or omit message payloads** unless an explicit, separately-gated flag is set.
- **Authentication / transport.** Support a token-gated stdio and/or HTTP transport; document running it co-located (same network/credentials as the app) rather than internet-exposed. Reuse the app's existing DB credentials/connection config rather than introducing a new privileged path.
- **Bounded queries.** Pagination and row caps on every tool (the DataSource methods already paginate) so a tool call can't run an unbounded scan on a large pgmq table.

## Packaging

- Ship as an optional component (e.g. `pgbus mcp` CLI subcommand, or a small companion that boots the MCP over the existing pgbus connection). Gate behind an explicit flag/env so it never starts unless asked.
- Document client setup (the tool list + a minimal config snippet) in the README.
- Provide a machine-readable `pgbus_health` verdict suitable for both interactive agent use and automated alerting.

## Acceptance criteria

- An MCP client can list and call the read-only tools above and get structured JSON.
- The diagnostic path for #179 is reproducible through the MCP alone: confirm backed-up queues, `read_ct = 0`, workers heart-beating, nothing paused/locked — no raw SQL.
- `pgbus_health` returns STALLED for the silent-wedge condition (backlog + idle-capacity worker + no read_ct progress) and OK when draining normally.
- No tool can mutate state; no tool returns message payloads unless an explicit, separately-gated flag is passed.

## Notes

- Reuse `lib/pgbus/web/data_source.rb` rather than writing new queries — it is the dashboard's proven read layer.
- Complements #179: that issue makes pgbus self-heal the wedge; this issue makes the wedge (and pgbus health generally) observable and triageable by agents/operators. Once #179 adds the loop-progress beacon and Stalled state, expose them through `pgbus_processes` / `pgbus_health` here.

### PR DESCRIPTION
**Pull request by mhenrixon**
## Summary

Adds an official, **read-only** [MCP](https://modelcontextprotocol.io) diagnostic server for pgbus (issue #180). An AI agent — or any MCP client — can now diagnose pgbus directly ("are queues backed up?", "is `read_ct` advancing?", "are workers heart-beating but not claiming?") instead of hand-writing `pgmq` / `pg_stat_activity` SQL against production.

Started only via `bundle exec pgbus mcp` (stdio). The 12 tools are a thin adapter over `lib/pgbus/web/data_source.rb` — the dashboard's proven read layer — so no new database access path is introduced.

The headline is `pgbus_health`, which returns an **OK / DEGRADED / STALLED** verdict. STALLED is the silent-worker-wedge signal (#179 / #174 / #181): visible backlog while workers heart-beat but never advance `read_ct`. It is suitable for both interactive agent use and automated alerting.

## Tools (all read-only)

| Tool | Maps to |
|------|---------|
| `pgbus_health` | `HealthAnalyzer` over `queues_with_metrics` + `processes` + `queue_health_stats` |
| `pgbus_queues` | `queues_with_metrics` |
| `pgbus_queue_detail` | `queue_detail` + `queue_paused?` + `queue_health_detail` |
| `pgbus_processes` | `processes` (incl. `healthy`/`stale`/`stalled`) |
| `pgbus_jobs` / `pgbus_job_detail` | `jobs` / `job_detail` |
| `pgbus_dlq` / `pgbus_dlq_detail` | `dlq_messages` / `dlq_message_detail` |
| `pgbus_locks` | `job_locks` |
| `pgbus_throughput` / `pgbus_stats` | `job_throughput` / `job_status_counts` + `job_stats_summary` |
| `pgbus_recurring` | `recurring_tasks` |

## Security (the "secure way" the issue asks for)

- **Read-only by contract.** `BaseTool` sets `read_only_hint`; there is no mutation tool and no raw-SQL passthrough — the tool set is fixed and curated.
- **Payloads redacted.** Message bodies / headers / job arguments become `[redacted]` unless **both** `PGBUS_MCP_ALLOW_PAYLOADS=1` and a per-call `include_payloads: true` are set (double gate).
- **Bounded queries.** List tools cap at 100 rows/page; time windows cap at 1440 minutes.
- **Reuses the app's DB credentials** — no new privileged path.
- **Optional token gate.** Set `PGBUS_MCP_TOKEN` and the server refuses to start unless `PGBUS_MCP_AUTH_TOKEN` matches (constant-time compare).

## Packaging

`mcp` is an **optional** dependency loaded lazily by `Pgbus::MCP.load!`. The subsystem is excluded from Zeitwerk (its `MCP::Tool` subclasses would otherwise reference the gem at autoload time), mirroring how vendor integrations are handled. Users who never run the server don't have to install the gem; the `pgbus mcp` command tells them how if it's missing.

## How this maps to the acceptance criteria

- ✅ An MCP client can list and call the read-only tools and get structured JSON (verified with a real `initialize` + `tools/list` stdio handshake).
- ✅ The #179 diagnostic path is reproducible through the MCP alone: `pgbus_queues` (backed up), `pgbus_jobs` (`read_ct = 0`), `pgbus_processes` (heart-beating), `pgbus_queue_detail` (nothing paused), `pgbus_locks` (nothing locked) — no raw SQL.
- ✅ `pgbus_health` returns STALLED for the silent-wedge condition and OK when draining normally.
- ✅ No tool mutates state; no tool returns payloads unless the explicit double gate is open.

## Test plan

- [x] `bundle exec rubocop lib/` — no offenses
- [x] `bundle exec rspec spec/pgbus/mcp spec/pgbus/cli_spec.rb` — 91 examples, 0 failures
- [x] Non-system suite: 2379 examples, 0 new failures (the 2 i18n failures pre-exist on `main`)
- [x] stdio `initialize` + `tools/list` handshake smoke test returns the server info and all 12 tools

Closes #180

## Summary by CodeRabbit

* **New Features**
  * Added an optional read-only MCP diagnostic server accessible from the CLI.
  * Introduced MCP tools for health, queues, jobs, DLQ, locks, throughput, recurring tasks, and detailed queue/job inspection.
  * Added payload redaction and optional payload visibility controls for diagnostics.

* **Documentation**
  * Updated the README with startup instructions, safety notes, and a sample client configuration.

* **Tests**
  * Expanded coverage for CLI startup, MCP server behavior, payload redaction, authorization, health analysis, and tool responses.

### CODE REVIEW

**Comment by mhenrixon** [comment]
@CodeRabbit review

**Comment by mhenrixon** [comment]
## Added: HTTP deployment via `Pgbus::MCP.rack_app` (8e3e6d0)

The original PR shipped stdio only (`pgbus mcp`), which is for a **local operator** whose MCP client spawns the process. That doesn't cover "deploy this as part of my Rails server" — so this commit adds the HTTP path.

**Mount it in the Rails app you already run — no second process:**

```ruby
# config/routes.rb
mount Pgbus::MCP.rack_app(token: ENV["PGBUS_MCP_TOKEN"]) => "/pgbus/mcp"
```

### Two deployment models, now both supported

| | stdio (`pgbus mcp`) | HTTP (`rack_app`) |
|---|---|---|
| Who connects | Local operator (Claude Desktop/Code) | Remote agent / alerting |
| Process | Client spawns it | Inside your Rails server |
| Use for | Interactive debugging | Automated/remote diagnostics |

### Why stateless + JSON-response mode

The transport runs `stateless: true, enable_json_response: true`, so every request is a self-contained POST returning one JSON object — **no in-memory session, no SSE, no reaper thread**. That makes it safe behind multiple Puma/Falcon workers (any worker answers any request), which the gem's default stateful mode is not. Read-only diagnostics never need server-initiated messages, which stateless disables anyway.

### Security

- `token:` → `Authorization: Bearer <token>`, constant-time compared, `401` + `WWW-Authenticate` on miss
- `auth:` → custom `->(rack_request)` callable (mirrors `config.web_auth`), wins over token
- Warns if mounted with neither
- Auth gate runs before `handle_request` for **every** HTTP method (POST/GET/DELETE), so the bare transport is never reachable unauthenticated
- Payload redaction enforced identically to stdio

### Validation

- 15 new specs (`rack_app_spec.rb`) driving a real `Rack::MockRequest`: 401/200/406 paths, custom auth, redaction over HTTP
- Design independently corroborated by a gem-source understanding sweep **and** an adversarial security review — no exploitable findings
- Full MCP suite: 106 examples, 0 failures

README updated with the two-model comparison, mount snippet, options table, client config for both transports, and a standalone-`config.ru` pod option.

File: lib/pgbus/mcp.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    # Raises Pgbus::Error with an actionable message if the gem is missing.
+    def load!
+      require "mcp"
+      REQUIRES.each { |path| require path }
+      true
+    rescue LoadError => e
+      raise e unless e.message.include?("mcp")
+
+      raise Pgbus::Error,
+            "The pgbus MCP server requires the `mcp` gem. Add `gem \"mcp\"` to your Gemfile and run `bundle install`."
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/mcp.rb]: Fixed in 02fe5ec3c032cc212411fffc6b3ea72c2d07b266 — only the bare `require "mcp"` is now wrapped in the LoadError rescue. A typo or missing file in REQUIRES surfaces its real LoadError instead of being mis-reported as a missing gem.

File: lib/pgbus/mcp/health_analyzer.rb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        }
+      end
+
+      private
+
+      def safe_queue_health
+        @data_source.queue_health_stats
+      rescue StandardError
+        {}
+      end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/mcp/health_analyzer.rb:79]: Fixed in 02fe5ec3c032cc212411fffc6b3ea72c2d07b266 — `safe_queue_health` now returns `[hash, error]`. The error is logged via `Pgbus.logger.warn` (per the no-silent-errors rule) and added to the DEGRADED reasons so the caller sees the verdict was built from partial data.

File: lib/pgbus/mcp/health_analyzer.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        live_workers    = workers.select { |w| %w[healthy stalled].include?(w[:status].to_s) }
+
+        reasons = []
+        if stalled_workers.any?
+          reasons << "#{stalled_workers.size} worker(s) stalled (heart-beating but claim loop not advancing) " \
+                     "while #{backlog.size} queue(s) have visible backlog: #{backlog_names(backlog)}"
+        elsif live_workers.any? && all_unread?(backlog)
+          reasons << "#{backlog.size} queue(s) have visible messages with read_ct=0 (never claimed) " \
+                     "while #{live_workers.size} worker(s) are alive: #{backlog_names(backlog)}"
+        end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/mcp/health_analyzer.rb]: Fixed in 02fe5ec3c032cc212411fffc6b3ea72c2d07b266 — `active_backlog = backlog.reject { |q| q[:paused] }` is what now feeds the STALLED check. A paused queue with read_ct=0 backlog correctly classifies as DEGRADED (paused-backlog), not STALLED. New spec at health_analyzer_spec.rb covers this.

File: lib/pgbus/mcp/tools/dlq_detail_tool.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+              description: "Include the raw message body/headers. Only honored when the server allows payloads."
+            }
+          },
+          required: %w[msg_id]
+        )
+
+        def self.call(msg_id:, include_payloads: false, server_context: nil)
+          data_source = data_source_from(server_context)
+          detail = data_source.dlq_message_detail(msg_id)
+          return error_response("Dead-letter message #{msg_id} not found") unless detail
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/mcp/tools/dlq_detail_tool.rb:48]: Fixed in 02fe5ec3c032cc212411fffc6b3ea72c2d07b266 — `DlqDetailTool` now accepts an optional `queue:` arg for direct lookup via `DataSource#job_detail`. Without it, `ambiguity_check` scans every DLQ and refuses to guess if the id is in more than one, returning an error response that asks the caller to pass `queue:`. Matches the `JobDetailTool` pattern.

File: lib/pgbus/mcp/tools/dlq_tool.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            }
+          },
+          required: []
+        )
+
+        def self.call(page: 1, per_page: 25, include_payloads: false, server_context: nil)
+          data_source = data_source_from(server_context)
+          per_page = per_page.to_i.clamp(1, MAX_PER_PAGE)
+          page = [page.to_i, 1].max
+          rows = data_source.dlq_messages(page: page, per_page: per_page)
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/mcp/tools/dlq_tool.rb:37]: Fixed in 02fe5ec3c032cc212411fffc6b3ea72c2d07b266 — added `MAX_PAGE = 1_000`; `page` is now `page.to_i.clamp(1, MAX_PAGE)` so OFFSET can't be driven to an unbounded DLQ scan. New spec verifies the cap.

File: lib/pgbus/mcp/tools/jobs_tool.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            }
+          },
+          required: []
+        )
+
+        def self.call(queue: nil, page: 1, per_page: 25, include_payloads: false, server_context: nil)
+          data_source = data_source_from(server_context)
+          per_page = per_page.to_i.clamp(1, MAX_PER_PAGE)
+          page = [page.to_i, 1].max
+          rows = data_source.jobs(queue_name: queue, page: page, per_page: per_page)
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/mcp/tools/jobs_tool.rb:45]: Fixed in 02fe5ec3c032cc212411fffc6b3ea72c2d07b266 — same MAX_PAGE=1000 cap applied to `JobsTool` so a large `page` can no longer drive deep OFFSET scans against the queue tables. New spec verifies the cap.

File: spec/pgbus/mcp/redactor_spec.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+RSpec.describe Pgbus::MCP::Redactor do
+  let(:row) do
+    {
+      msg_id: 42,
+      read_ct: 1,
+      queue_name: "pgbus_default",
+      message: "{\"secret\":\"hunter2\"}",
+      headers: "{\"trace\":\"abc\"}"
+    }
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/pgbus/mcp/redactor_spec.rb:14]: Pushing back on this one. The repo rule about hardcoded queue names is about *production code* using prefixed names via `config.queue_name()` — the alternative there would be a literal that drifts from a custom `queue_prefix`. In this spec `pgbus_default` is opaque fixture data: the redactor doesn't care what the queue is named, it only redacts payload keys. Replacing it with `Pgbus.configuration.queue_name(:default)` would couple a payload-redaction unit test to the queue-name builder for no extra signal. Happy to revisit if you can show a real drift path.
